packages/docs-page/server/generate-static-props.ts

```diff
@@ -0,0 +1,154 @@
+import moize, { Options } from 'moize'
+
+import {
+  fetchNavData,
+  fetchDocument,
+  fetchVersionMetadataList,
+} from '../content-api'
+import { ENABLE_VERSIONED_DOCS, DEFAULT_PARAM_ID } from './consts'
+
+import renderPageMdx from '../render-page-mdx'
+
+import { stripVersionFromPathParams, normalizeVersion } from '../util'
+
+import { resolveNavData } from './resolve-nav-data'
+
+const moiseOpts: Options = { isPromise: true, maxSize: Infinity }
+const cachedFetchNavData = moize(fetchNavData, moiseOpts)
+const cachedFetchVersionMetadataList = moize(
+  fetchVersionMetadataList,
+  moiseOpts
+)
+
+export async function generateStaticProps({
+  navDataFile,
+  localContentDir,
+  params, // ["destroy"] | ["v0.5.x", "destroy"]
+  product: { name: productName, slug: productSlug },
+  mainBranch = 'main',
+  remarkPlugins = [],
+  scope, // optional, I think?
+  paramId = DEFAULT_PARAM_ID,
+  basePath,
+}) {
+  const mdxRenderer = (mdx) =>
+    renderPageMdx(mdx, {
+      productName,
+      remarkPlugins,
+      scope,
+    })
+
+  // Build the currentPath from page parameters
+  const currentPath = params[paramId] ? params[paramId].join('/') : ''
+
+  let versions = []
+  let latestVersion = null
+
+  // This code path handles versioned docs integration, which is currently gated behind the ENABLE_VERSIONED_DOCS env var
+  if (ENABLE_VERSIONED_DOCS) {
+    // given: v0.5.x (latest), v0.4.x, v0.3.x
+    const [versionFromPath, paramsNoVersion] = stripVersionFromPathParams(
+      params[paramId]
+    )
+    // versionFromPath should realistically only ever be "latest" | "v0.4.x" | "v0.3.x"
+    // It could be v0.5.x if a use navigates directly to it.
+
+    const versionMetadataList = await cachedFetchVersionMetadataList(
+      productSlug
+    )
+    versions = versionMetadataList.map((e) => {
+      const { isLatest, version, display } = e
+      if (isLatest) {
+        latestVersion = version
+      }
+
+      const displayValue = display ?? version
+
+      return {
+        name: isLatest ? 'latest' : version,
+        label: isLatest ? `${displayValue} (latest)` : displayValue,
+      }
+    })
```

Comment:
seems like this could be a fair candidate to extract into a function that's easier to test

Reply:
extracted into `mapVersionList`